src/Illuminate/Routing/UrlGenerator.php

```diff
@@ -359,15 +360,26 @@ protected function formatAction($action)
      * Format the array of URL parameters.
      *
      * @param  mixed|array  $parameters
+     * @param  Route        $route
      * @return array
      */
-    public function formatParameters($parameters)
+    public function formatParameters($parameters, Route $route = null)
     {
         $parameters = array_wrap($parameters);
+        $routeParameters = $route ? $route->parameterNames() : [];
 
-        foreach ($parameters as $key => $parameter) {
+        foreach ($parameters as $index => $parameter) {
+            if (isset($routeParameters[$index])) {
+                $key = $routeParameters[$index];
+            }
+            $parts = explode(':', isset($key) ? $key : $index);
+            $key = isset($parts[1]) ? $parts[1] : null;
             if ($parameter instanceof UrlRoutable) {
-                $parameters[$key] = $parameter->getRouteKey();
+                if ($parameter instanceof Model) {
```

Comment:
I agree and I changed it, but for the record: it's also being used here: https://github.com/laravel/framework/blob/5.4/src/Illuminate/Routing/ImplicitRouteBinding.php#L5
`illuminate/database` is not in the composer.json though...